python/pdo/contract/replication.py

```diff
@@ -0,0 +1,277 @@
+                break
+
+    except Exception as e:
+        logger.info("Replication Worker exception %s", str(e))
+
+# -----------------------------------------------------------------
+class ReplicationRequest(object):
+    """ implements the replicator class functionality : used for change-set replication after
+    contract update, before commiting transaction."""
+
+    # -----------------------------------------------------------------
+    def __init__(self, replication_params, \
+                contract_id, blocks_to_replicate, commit_id, data_dir=None):
+        """ Create an instance of the ReplicationRequest class: used for replicating the current change set.
+        eservice_ids can be replaced with sservice_ids after we create an sservice database that can be used to look up the sservice url using the id."""
+
+        self.service_ids = replication_params['service_ids']
+        self.num_provable_replicas = replication_params['num_provable_replicas']
+        self.availability_duration = replication_params['availability_duration']
+        self.contract_id = contract_id
+        self.blocks_to_replicate = blocks_to_replicate
+        self.commit_id = commit_id
+        self.data_dir = data_dir
+
+        self.is_completed = False
+        self.is_failed = False
+        self.successful_services = set()
+        self.unsuccessful_services = set()
+
+    # -----------------------------------------------------------------
+    def mark_as_completed(self, call_back_after_replication=None):
+        """ Mark as completed. Notify waiting threads. If successful, invoke the call back method. Multiple notifications and call backs are prevented"""
+
+        if not self.is_completed: # check reduces contention for lock
+            __condition_variable_for_completed_tasks__.acquire()
+            if not self.is_completed: # yes we check this again, the outside check is only for performance optimization, this check is algorithmic
+                self.is_completed = True
+                logger.info("Replication for request number %d successfully completed", self.commit_id[2])
+                __condition_variable_for_completed_tasks__.notify()
+                if not self.is_failed:
+                    call_back_after_replication()
+            __condition_variable_for_completed_tasks__.release()
+
+    # -----------------------------------------------------------------
+    def mark_as_failed(self):
+        self.is_failed = True
+        self.mark_as_completed()
+
+    # -----------------------------------------------------------------
+    def wait_for_completion(self):
+        """ Returns after successful completion of the replication request. Raises exception if the request failed."""
+
+        release_lock = False
+        while self.is_completed is False:
+            __condition_variable_for_completed_tasks__.acquire()
+            __condition_variable_for_completed_tasks__.wait()
+            release_lock = True
```

Comment:
further simplification:
now that there is no timeout, the code will always check for completion, and acquire and release the lock so... you don't need the `release_lock` variable and check anymore

Reply:
thank you, simplified the code in the latest pull.